internal/configs/config.go

```diff
@@ -487,45 +522,75 @@ func (c *Config) addProviderRequirements(reqs getproviders.Requirements, recurse
 
 	// "provider" block can also contain version constraints
 	for _, provider := range c.Module.ProviderConfigs {
-		fqn := c.Module.ProviderForLocalConfig(addrs.LocalProviderConfig{LocalName: provider.Name})
-		if _, ok := reqs[fqn]; !ok {
-			// We'll at least have an unconstrained dependency then, but might
-			// add to this in the loop below.
-			reqs[fqn] = nil
-		}
-		if provider.Version.Required != nil {
-			// The model of version constraints in this package is still the
-			// old one using a different upstream module to represent versions,
-			// so we'll need to shim that out here for now. The two parsers
-			// don't exactly agree in practice 🙄 so this might produce new errors.
-			// TODO: Use the new parser throughout this package so we can get the
-			// better error messages it produces in more situations.
-			constraints, err := getproviders.ParseVersionConstraints(provider.Version.Required.String())
-			if err != nil {
-				diags = diags.Append(&hcl.Diagnostic{
-					Severity: hcl.DiagError,
-					Summary:  "Invalid version constraint",
-					// The errors returned by ParseVersionConstraint already include
-					// the section of input that was incorrect, so we don't need to
-					// include that here.
-					Detail:  fmt.Sprintf("Incorrect version constraint syntax: %s.", err.Error()),
-					Subject: provider.Version.DeclRange.Ptr(),
-				})
+		moreDiags := c.addProviderRequirementsFromProviderBlock(reqs, provider)
+		diags = append(diags, moreDiags...)
+	}
+
+	// We may have provider blocks and required_providers set in some testing
+	// files.
+	if tests {
```

Comment:
I'm not quite sure I understand this, shouldn't the requirements be the same regardless of whether there are providers configured within tests?

Reply:
I don't think you can have different provider requirements. We need to have a stable set defined in the provider lock file, plus you wouldn't want the tests running with a different provider version invalidating the test results. There could be _additional_ provider defines in the test block, but those need to be resolved already. (from the Go perspective, `go.mod` must always accounts for test dependencies too, and you wouldn't run `go test` with a different `go.mod`)

Reply:
Yeah, all of this is being read and considered during `terraform init`. 

So any provider requirements set in the test dependencies (whether in dedicated test modules or provider blocks) will be checked, and additional providers downloaded, when the user runs `terraform init`. It's all just extra requirements that the provider fetcher needs to take into account when getting the providers and building the lock file.

There won't be different requirements, it's just additional requirements that need to be accounted for. If the tests say I need a provider of version x and the configuration says I need a provider of version y then the init command will fail and say I can't find a provider with a valid version, as it does at the moment if you included two modules in your configuration with incompatible requirements.

Hopefully that makes sense, happy to have a quick call to discuss if needed!

Reply:
In general, users shouldn't be setting provider versions directly in the test files (as they shouldn't be in real configuration as well) as that attribute is deprecated. 

But they can write modules that are only loaded by the tests that might pull in providers that aren't needed by the main configuration, so those modules need to be processed by `terraform init` for provider requirements, but if those requirements are incompatible with the config under test then that's a problem for the user to fix, and init will tell them they have incompatible version requirements.

Reply:
Thanks! I think I had a false assumption about the ordering of operations in my head last week. As long as all provider requirements are always accounted for this looks good.